fairmq/sdk/top/main.cxx

```diff
@@ -0,0 +1,142 @@
+}
+
+template<typename Executor>
+auto makeTop(Executor ex, fair::mq::sdk::DDSSession::Id sessionId) -> fair::mq::sdk::Top
+{
+    using namespace fair::mq;
+
+    if (sessionId.empty()) {
+        return sdk::Top(std::move(ex), sdk::getMostRecentRunningDDSSession());
+    }
+    return sdk::Top(std::move(ex), sdk::DDSSession(sessionId));
+}
+
+struct NoCopy
+{
+    NoCopy(const NoCopy&) = delete;
+    NoCopy& operator=(const NoCopy&) = delete;
+
+    NoCopy() = default;
+    NoCopy(NoCopy&&) = default;
+    NoCopy& operator=(NoCopy&&) = default;
+    ~NoCopy() = default;
+};
+
+} // anonymous namespace
+
+int main(int argc, char const* argv[])
+{
+    using namespace boost;
+    using namespace fair::mq;
+
+    sdk::DDSSession::Id sessionId;
+    auto sessionOptionValue(program_options::value<sdk::DDSSession::Id>(&sessionId));
+    auto envSessionId = getenv("DDS_SESSION_ID");
+    if (envSessionId != nullptr) {
+        sessionOptionValue->default_value(sdk::DDSSession::Id(envSessionId));
+    }
+
+    program_options::options_description options("Options");
+    options.add_options()
+        ("session,s", sessionOptionValue,
+                      "DDS Session ID (overrides $DDS_SESSION_ID). If no session id is "
+                      "provided it tries to find most recent running session.")
+        ("help,h", "Print usage help");
+
+    try {
+        program_options::variables_map vm;
+        program_options::store(
+            program_options::command_line_parser(argc, argv).options(options).run(), vm);
+        program_options::notify(vm);
+
+        if (vm.count("help") > 0) {
+            printHelp(options);
+            return EXIT_SUCCESS;
+        }
+
+        // Execution context
+        asio::io_context context;
+
+        // Abort on SIGINT and SIGTERM
+        asio::signal_set signals(context.get_executor(), SIGINT, SIGTERM);
+        signals.async_wait([&](const system::error_code&, int) { context.stop(); });
+
+        // Run fairmq-top
+        auto app = makeTop(context.get_executor(), sessionId);
+        int rc(EXIT_SUCCESS);
+
+        ///// move-only lambda
+        //
+        NoCopy noCopy; // just to check completion lambda is never copied
+        app.AsyncRun([&, nc=std::move(noCopy)](const system::error_code& ec) {
+            if (ec) {
+                rc = EXIT_FAILURE;
+            } else {
+                rc = EXIT_SUCCESS;
+            }
+            context.stop();
+        });
+
+        context.run();
+        //
+        /////////////////////
+
+        /// OR
+
+        ///// std::future
+        //
+        // std::thread ctxThread([&]{ context.run(); });
+
+        // auto future = app.AsyncRun(asio::use_future);
+
+        // try {
+            // future.get();
+            // rc = EXIT_SUCCESS;
+        // } catch (const system::error_code& e) {
+            // rc = EXIT_FAILURE;
+        // }
+
+        // context.stop();
+        // ctxThread.join();
+        //
```

Comment:
@mattkretz Trying to comply with asio's [AsynchronousOperation](https://www.boost.org/doc/libs/1_70_0/doc/html/boost_asio/reference/asynchronous_operations.html) concept. Calling with a completion token in the form of a lambda, [my code](https://github.com/FairRootGroup/FairMQ/pull/195/files#diff-3dfc84dcdb355ed30b83b782ee6d39e0R104) compiles and works as expected. Using the `boost::asio::use_future` completion token instead like this (in https://github.com/FairRootGroup/FairMQ/pull/195/files#diff-3dfc84dcdb355ed30b83b782ee6d39e0R123):

```cpp
        std::thread ctxThread([&]{ context.run(); });

        auto future = app.AsyncRun(asio::use_future);

        try {
            future.get();
            rc = EXIT_SUCCESS;
        } catch (const system::error_code& e) {
            rc = EXIT_FAILURE;
        }

        context.stop();
        ctxThread.join();
```

leads to the following compiler error:

```
[1/3] Building CXX object fairmq/sdk/top/CMakeFiles/fairmq-top.dir/main.cxx.o
FAILED: fairmq/sdk/top/CMakeFiles/fairmq-top.dir/main.cxx.o
/usr/bin/c++  -DBOOST_ERROR_CODE_HEADER_ONLY -DFAIRLOGGER_USE_BOOST_PRETTY_FUNCTION -I../ -I. -isystem /home/dklein/boost_1_70_0_install/include -isystem /home/dklein/projects/FairLogger/build/install/include -fdiagnostics-color=always -O2 -g -Wshadow -Wall -Wextra -DNDEBUG -fPIE   -pthread -D_DEFAULT_SOURCE -D_XOPEN_SOURCE=600 -std=c++14 -MD -MT fairmq/sdk/top/CMakeFiles/fairmq-top.dir/main.cxx.o -MF fairmq/sdk/top/CMakeFiles/fairmq-top.dir/main.cxx.o.d -o fairmq/sdk/top/CMakeFiles/fairmq-top.dir/main.cxx.o -c ../fairmq/sdk/top/main.cxx
In file included from ../fairmq/sdk/top/main.cxx:15:
../fairmq/sdk/top/Top.h: In instantiation of ‘fair::mq::sdk::BasicTop<Executor>::AsyncRun(CompletionToken&&) [with CompletionToken = const boost::asio::use_future_t<>&; Executor = boost::asio::executor]::<lambda(auto:1)> [with auto:1 = boost::asio::detail::promise_handler<void(boost::system::error_code), std::allocator<void> >]’:
/home/dklein/boost_1_70_0_install/include/boost/asio/async_result.hpp:273:47:   required from ‘typename std::enable_if<(! boost::asio::detail::async_result_has_initiate_memfn<CompletionToken, Signature>::value), typename boost::asio::async_result<typename std::decay<_Tp>::type, Signature>::return_type>::type boost::asio::async_initiate(Initiation&&, CompletionToken&, Args&& ...) [with CompletionToken = const boost::asio::use_future_t<>&; Signature = void(boost::system::error_code); Initiation = fair::mq::sdk::BasicTop<Executor>::AsyncRun(CompletionToken&&) [with CompletionToken = const boost::asio::use_future_t<>&; Executor = boost::asio::executor]::<lambda(auto:1)>; Args = {}; typename std::enable_if<(! boost::asio::detail::async_result_has_initiate_memfn<CompletionToken, Signature>::value), typename boost::asio::async_result<typename std::decay<_Tp>::type, Signature>::return_type>::type = std::future<void>]’
../fairmq/sdk/top/Top.h:86:79:   required from ‘auto fair::mq::sdk::BasicTop<Executor>::AsyncRun(CompletionToken&&) [with CompletionToken = const boost::asio::use_future_t<>&; Executor = boost::asio::executor]’
../fairmq/sdk/top/main.cxx:123:52:   required from here
../fairmq/sdk/top/Top.h:93:26: error: no match for call to ‘(const boost::asio::detail::promise_handler<void(boost::system::error_code), std::allocator<void> >) (boost::system::error_code)’
   93 |                         h(system::errc::make_error_code(system::errc::success));
      |                         ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/dklein/boost_1_70_0_install/include/boost/asio/use_future.hpp:157,
                 from ../fairmq/sdk/top/main.cxx:13:
/home/dklein/boost_1_70_0_install/include/boost/asio/impl/use_future.hpp:303:8: note: candidate: ‘void boost::asio::detail::promise_handler_ec_0::operator()(const boost::system::error_code&)’ <near match>
  303 |   void operator()(const boost::system::error_code& ec)
      |        ^~~~~~~~
/home/dklein/boost_1_70_0_install/include/boost/asio/impl/use_future.hpp:303:8: note:   passing ‘const boost::asio::detail::promise_handler<void(boost::system::error_code), std::allocator<void> >*’ as ‘this’ argument discards qualifiers
/home/dklein/boost_1_70_0_install/include/boost/asio/impl/use_future.hpp:228:8: error: ‘void boost::asio::detail::promise_executor<T>::post(F&&, const A&) const [with F = fair::mq::sdk::BasicTop<Executor>::AsyncRun(CompletionToken&&) [with CompletionToken = const boost::asio::use_future_t<>&; Executor = boost::asio::executor]::<lambda(auto:1)> [with auto:1 = boost::asio::detail::promise_handler<void(boost::system::error_code), std::allocator<void> >]::<lambda()>; A = std::allocator<void>; T = void]’, declared using local type ‘fair::mq::sdk::BasicTop<Executor>::AsyncRun(CompletionToken&&) [with CompletionToken = const boost::asio::use_future_t<>&; Executor = boost::asio::executor]::<lambda(auto:1)> [with auto:1 = boost::asio::detail::promise_handler<void(boost::system::error_code), std::allocator<void> >]::<lambda()>’, is used but never defined [-fpermissive]
  228 |   void post(BOOST_ASIO_MOVE_ARG(F) f, const A& a) const
      |        ^~~~
ninja: build stopped: subcommand failed.
```

The mentioned `note: candidate: ‘void boost::asio::detail::promise_handler_ec_0::operator()(const boost::system::error_code&)’ <near match>` is what I expected to be available, but something not obvious to me is preventing my implementation from choosing it. You can find some of the relevant templates here:
* `../fairmq/sdk/top/Top.h:93` [`fair::mq::sdk::Top::AsyncRun`](https://github.com/FairRootGroup/FairMQ/pull/195/files#diff-556ca1f50fc2fcbd48924322db6e606cR93)
* [`boost::asio::detail::promise_handler`](https://github.com/boostorg/asio/blob/boost-1.70.0/include/boost/asio/impl/use_future.hpp#L589) and its base classes are defined in the same file above.

Would be awesome if you could have a quick look, thx!

Reply:
Implementing `sdk::Top::AsyncRun` without the `boost::asio::async_initiate` helper like this

```cpp
    template<typename CompletionToken>
    auto AsyncRun(CompletionToken&& token)
    {
        using namespace boost;

        asio::async_completion<CompletionToken, void(system::error_code)> completion(token);

        auto ex(asio::get_associated_executor(completion.completion_handler, fExecutor));
        auto alloc(asio::get_associated_allocator(completion.completion_handler));

        ex.post(
            [h = std::move(completion.completion_handler)]() {
                h(system::errc::make_error_code(system::errc::success));
            },
            alloc);

        return completion.result.get();

        // return asio::async_initiate<CompletionToken, void(system::error_code)>(
            // [&](auto handler) {
                // auto ex(asio::get_associated_executor(handler, fExecutor));
                // auto alloc(asio::get_associated_allocator(handler));
//
                // ex.post(
                    // [h = std::move(handler)]() {
                        // h(system::errc::make_error_code(system::errc::success));
                    // },
                    // alloc);
            // },
            // token);
    }
```

condenses the compiler error message a bit:

```
[1/3] Building CXX object fairmq/sdk/top/CMakeFiles/fairmq-top.dir/main.cxx.o
FAILED: fairmq/sdk/top/CMakeFiles/fairmq-top.dir/main.cxx.o
/usr/bin/c++  -DBOOST_ERROR_CODE_HEADER_ONLY -DFAIRLOGGER_USE_BOOST_PRETTY_FUNCTION -I../ -I. -isystem /home/dklein/boost_1_70_0_install/include -isystem /home/dklein/projects/FairLogger/build/install/include -fdiagnostics-color=always -O2 -g -Wshadow -Wall -Wextra -DNDEBUG -fPIE   -pthread -D_DEFAULT_SOURCE -D_XOPEN_SOURCE=600 -std=c++14 -MD -MT fairmq/sdk/top/CMakeFiles/fairmq-top.dir/main.cxx.o -MF fairmq/sdk/top/CMakeFiles/fairmq-top.dir/main.cxx.o.d -o fairmq/sdk/top/CMakeFiles/fairmq-top.dir/main.cxx.o -c ../fairmq/sdk/top/main.cxx
In file included from ../fairmq/sdk/top/main.cxx:15:
../fairmq/sdk/top/Top.h: In instantiation of ‘auto fair::mq::sdk::BasicTop<Executor>::AsyncRun(CompletionToken&&) [with CompletionToken = const boost::asio::use_future_t<>&; Executor = boost::asio::executor]’:
../fairmq/sdk/top/main.cxx:123:52:   required from here
../fairmq/sdk/top/Top.h:93:18: error: no match for call to ‘(const boost::asio::detail::promise_handler<void(boost::system::error_code), std::allocator<void> >) (boost::system::error_code)’
   93 |                 h(system::errc::make_error_code(system::errc::success));
      |                 ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/dklein/boost_1_70_0_install/include/boost/asio/use_future.hpp:157,
                 from ../fairmq/sdk/top/main.cxx:13:
/home/dklein/boost_1_70_0_install/include/boost/asio/impl/use_future.hpp:303:8: note: candidate: ‘void boost::asio::detail::promise_handler_ec_0::operator()(const boost::system::error_code&)’ <near match>
  303 |   void operator()(const boost::system::error_code& ec)
      |        ^~~~~~~~
/home/dklein/boost_1_70_0_install/include/boost/asio/impl/use_future.hpp:303:8: note:   passing ‘const boost::asio::detail::promise_handler<void(boost::system::error_code), std::allocator<void> >*’ as ‘this’ argument discards qualifiers
ninja: build stopped: subcommand failed.
```

Reply:
your `operator()` is missing a const qualification. i.e. `operator()(...) const`

Reply:
ah well, it's asio's `operator()` that's non-const. So your problem is that `promise_handler` must be non-const.

Reply:
thx!